runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,1110 @@
+	/// BlockNumber.
+	last_slot: BlockNumber,
+	/// The deployment data associated with this fund, if any. Once set it may not be reset. First
+	/// is the code hash, second is the initial head data.
+	deploy_data: Option<(Hash, Vec<u8>)>,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		/// Info on all of the funds.
+		Funds get(funds):
+			map FundIndex => Option<FundInfo<T::AccountId, BalanceOf<T>, T::Hash, T::BlockNumber, ParaIdOf<T>>>;
+
+		/// The total number of funds that have so far been allocated.
+		FundCount get(fund_count): FundIndex;
+
+		/// The funds that have had additional contributions during the last block. This is used
+		/// in order to determine which funds should submit new or updated bids.
+		NewRaise get(new_raise): Vec<FundIndex>;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as system::Trait>::AccountId,
+		Balance = BalanceOf<T>,
+		ParaId = ParaIdOf<T>,
+	{
+		Created(FundIndex),
+		Contributed(AccountId, FundIndex, Balance),
+		Withdrew(AccountId, FundIndex, Balance),
+		Retiring(FundIndex),
+		Dissolved(FundIndex),
+		DeployDataFixed(FundIndex),
+		Onboarded(FundIndex, ParaId),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+		
+		/// Create a new crowdfunding campaign for a parachain slot deposit for the current auction.
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000)]
+		fn create(origin,
+			#[compact] cap: BalanceOf<T>,
+			#[compact] first_slot: T::BlockNumber,
+			#[compact] last_slot: T::BlockNumber
+		) {
+			let owner = ensure_signed(origin)?;
+
+			ensure!(first_slot < last_slot, "last slot must be greater than first slot");
+			ensure!(last_slot <= first_slot + 3.into(), "last slot cannot be more then 3 more than first slot");
+			// Check an auction is in progress, and extract the `early_end` block
+			let (_, early_end) = <slots::Module<T>>::auction_info().ok_or("no auction in progress")?;
+
+			// End of the crowdfund will be the last possible block for the ongoing auction
+			let end = early_end + T::EndingPeriod::get();
+
+			let deposit = T::SubmissionDeposit::get();
+			let imb = T::Currency::withdraw(
+				&owner,
+				deposit,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath,
+			)?;
+
+			let index = FundCount::get();
+			let next_index = index.checked_add(1).ok_or("overflow when adding fund")?;
+			FundCount::put(next_index);
+
+			// No fees are paid here if we need to create this account; that's why we don't just
+			// use the stock `transfer`.
+			T::Currency::resolve_creating(&Self::fund_account_id(index), imb);
+
+			<Funds<T>>::insert(index, FundInfo {
+				parachain: None,
+				owner: owner,
+				deposit: deposit,
+				raised: Zero::zero(),
+				end: end,
+				cap: cap,
+				last_contribution: None,
+				first_slot: first_slot,
+				last_slot: last_slot,
```

Comment:
these `name: name,` lines can be omitted with a `..` at the end. less clutter -> good.